File: nexus-backend/src/main/java/tn/arctic/nexus/services/JamsModule/EntryRatingService.java
package tn.arctic.nexus.services.JamsModule;

import tn.arctic.nexus.entities.EntryMedia;
import tn.arctic.nexus.entities.EntryRating;
import tn.arctic.nexus.repositories.JamsModule.IEntryRatingRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class EntryRatingService implements IEntryRatingService {

    @Autowired
    private IEntryRatingRepository ratingRepository;

    @Override
    public List<EntryRating> getAllRatings() {
        return ratingRepository.findAll();
    }

    @Override
    public List<EntryRating> getRatingsByEntry(Long entryId) {
        return ratingRepository.findByEntryId(entryId);
    }

    @Override
    public Optional<EntryRating> getRatingById(Long id) {
        return ratingRepository.findById(id);
    }

    @Override
    public EntryRating createRating(EntryRating rating) {
        boolean alreadyRated = ratingRepository
                .findByEntryIdAndUserId(rating.getEntry().getId(), rating.getUser().getId())
                .isPresent();

        if (alreadyRated) {
            throw new RuntimeException("You have already rated this entry.");
        }

        return ratingRepository.save(rating);
    }


    @Override
    public EntryRating updateRating(EntryRating entryRating) {
        return ratingRepository.save(entryRating);
    }
    @Override
    public void deleteRating(Long id) {
        ratingRepository.deleteById(id);
    }



    @Override
    public boolean hasUserRatedEntry(Long userId, Long entryId) {
        return ratingRepository.findByEntryIdAndUserId(entryId, userId).isPresent();
    }

}

File: nexus-frontend/src/app/core/services/jam/entry-rating.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { EntryRating } from '../../entities/Jam/entry-rating';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class EntryRatingService {
  private baseUrl = '/api/nexus-backend/api/entry-ratings';

  constructor(private http: HttpClient) {}

  getAllRatings(): Observable<EntryRating[]> {
    return this.http.get<EntryRating[]>(`${this.baseUrl}/all`);
  }

  getRatingById(id: number): Observable<EntryRating> {
    return this.http.get<EntryRating>(`${this.baseUrl}/${id}`);
  }

  createRating(rating: EntryRating): Observable<EntryRating> {
    return this.http.post<EntryRating>(`${this.baseUrl}/add`, rating);
  }

  updateRating(rating: EntryRating): Observable<EntryRating> {
    return this.http.put<EntryRating>(`${this.baseUrl}/update`, rating);
  }

  deleteRating(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/delete/${id}`);
  }

  getRatingsByEntry(entryId: number): Observable<EntryRating[]> {
    return this.http.get<EntryRating[]>(`${this.baseUrl}/entry/${entryId}`);
  }

hasUserRated(entryId: number, userId: number): Observable<boolean> {
  return this.http.get<boolean>(`/api/nexus-backend/api/entry-ratings/has-rated/${entryId}/${userId}`);
}

  
}
